python/ray/data/_internal/table_block.py

```diff
@@ -206,6 +207,14 @@ def _zip(self, acc: BlockAccessor) -> "Block":
     def zip(self, other: "Block") -> "Block":
         acc = BlockAccessor.for_block(other)
         if not isinstance(acc, type(self)):
+            pd = _lazy_import_pandas()
+            if (
+                pd
+                and isinstance(self._table, pd.DataFrame)
+                and hasattr(other, "to_pandas")
+            ):
+                return self.zip(other.to_pandas())
```

Comment:
can you clarify why this code path should be added?

Reply:
Here I assumed `pandas` to be the fallback so that we only convert the `other` block `to_pandas` if the table has `pandas` blocks.

If we run into the scenario where the table has `arrow` blocks, the swap happens in the `ZipOperator`. Zipping between two `arrow` blocks is unaffected by this change.

Reply:
Without this logic this test case fails:
```
    def test_zip_datasets_with_different_types(ray_start_regular_shared):
        df = pd.DataFrame({"spam": [0]})
        ds1 = ray.data.from_pandas(df)
        ds2 = ray.data.from_items([{"ham": [0]}])
        zipped_ds = ds1.zip(ds2)
>       assert list(zipped_ds.iter_rows()) == [{"spam": 0, "ham": 0}]
```

Stacktrace:
```
python/ray/data/tests/test_formats.py:329: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
python/ray/data/iterator.py:271: in _wrapped_iterator
    for batch in batch_iterable:
python/ray/data/iterator.py:181: in _create_iterator
    block_iterator, stats, blocks_owned_by_consumer = self._to_block_iterator()
python/ray/data/_internal/iterator/iterator_impl.py:32: in _to_block_iterator
    block_iterator, stats, executor = ds._plan.execute_to_iterator()
python/ray/data/_internal/plan.py:539: in execute_to_iterator
    block_iter = itertools.chain([next(gen)], gen)
python/ray/data/_internal/execution/legacy_compat.py:54: in execute_to_legacy_block_iterator
    for bundle in bundle_iter:
python/ray/data/_internal/execution/interfaces/executor.py:37: in __next__
    return self.get_next()
python/ray/data/_internal/execution/streaming_executor.py:129: in get_next
    raise item
python/ray/data/_internal/execution/streaming_executor.py:187: in run
    while self._scheduling_loop_step(self._topology) and not self._shutdown:
python/ray/data/_internal/execution/streaming_executor.py:267: in _scheduling_loop_step
    update_operator_states(topology)
python/ray/data/_internal/execution/streaming_executor_state.py:359: in update_operator_states
    op.all_inputs_done()
python/ray/data/_internal/execution/operators/zip_operator.py:65: in all_inputs_done
    self._output_buffer, self._stats = self._zip(
python/ray/data/_internal/execution/operators/zip_operator.py:178: in _zip
    output_metadata = ray.get(output_metadata)
python/ray/_private/auto_init_hook.py:24: in auto_init_wrapper
    return fn(*args, **kwargs)
python/ray/_private/client_mode_hook.py:103: in wrapper
    return func(*args, **kwargs)
    
>                       raise value.as_instanceof_cause()
E                       ray.exceptions.RayTaskError(ValueError): ray::_zip_one_block() (pid=73167, ip=127.0.0.1)
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/execution/operators/zip_operator.py", line 241, in _zip_one_block
E                           result = _try_zip(block, other_block)
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/execution/operators/zip_operator.py", line 264, in _try_zip
E                           raise e
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/execution/operators/zip_operator.py", line 253, in _try_zip
E                           return BlockAccessor.for_block(block).zip(other_block)
E                         File "/Users/michaelhly/Projects/ray/python/ray/data/_internal/table_block.py", line 210, in zip
E                           raise ValueError(
E                       ValueError: Cannot zip <class 'ray.data._internal.pandas_block.PandasBlockAccessor'> with block of type <class 'pyarrow.lib.Table'>
```

Reply:
Actually, nvm. Your way works.